include/ttmlir/Dialect/TTNN/IR/TTNNOps.td

```diff
@@ -426,7 +426,11 @@ def TTNN_EmptyOp : TTNN_Op<"empty"> {
       Tensor empty operation
     }];
 
-    let arguments = (ins TT_Device:$device);
+    let arguments = (ins Optional<TT_Device>:$device,
+                         TTNN_ShapeAttr:$shape,
+                         OptionalAttr<TT_DataTypeAttr>:$dtype,
+                         OptionalAttr<TTNN_LayoutAttr>:$layout,
+                         OptionalAttr<TTNN_MemoryConfigAttr>:$memory_config);
```

Comment:
Can't all of these be inferred from the result type? Otherwise we have to assert they are in agreement.

Reply:
I gave this some thought before I typed it up... Yes, we can infer all the info from output, but it's not easy to work with and loses info. I prefer that we be explicit as to what the method itself is trying to achieve. E.g. calling `to_layout` with only `TensorLayout::Tile` makes it obvious what the function call is trying to achieve.

A good supporting example I had in mind was: imagine having a reshape op, you can infer what the intended shape from the output tensor's layout attr, but you'd still want to supply the input args to the function in the IR.

This approach also makes it much easier to work with emitc and runtime.

> Otherwise we have to assert they are in agreement.

Yes, agreed! I'll add verifiers for all the ops.

Reply:
Yeah let's add verifiers for the ops